### ISSUE TITLE
Multiple Selection field is not capturing user's choice

### ISSUE THREAD

**Issue by Brada65** [kordan/moodle-mod_surveypro#341]
We are using SurveyPro on our production site, and one of colleague find that multiple selection field is just exporting first option from the list, doesn't show real user selection.

**Comment by Brada65** [comment]
Just to clarify, when we go to look form (edit user's answers) we can see real choices (multiple), but when we try to export to pdf or xls it's exporting just first or first two options. Can you please look at this.

**Comment by kordan** [comment]
Ciao Brada65 and thanks for this report.
I tested your issue but I can't reproduce it. Her the export works fine for multiselects too.
I am using M32 with last surveypro from master WITH a lot of PR merged locally.
The PR I merged locally (waiting for official merge) over surveypro master head are:

git merge wrong_groupby_frequency
git merge filter_actually_enrolled
git merge centralizelinks
git merge filtercsvimport
git merge github_322
git merge notrim_label
git merge added_select_id
git merge bulk_action_help
git merge drop_chained_steps
git merge undefined_fillinginstruction
git merge get_dummy_contents_uses_bad_names
git merge shorter_submission_test

Are we using the same code?

**Comment by Brada65** [comment]
[SurveyPro error example of multi select field.pdf](https://github.com/kordan/moodle-mod_surveypro/files/962516/SurveyPro.error.example.of.multi.select.field.pdf)
We are using Moodle 3.2.1, and I installed new master with all PR merged locally, only thing that I can think of is that you send me your folder zipped (from your working Moodle), to try that one.

**Comment by meesir** [comment]
I just did some testing and am having the same or a similar issue.  Multi-select items are recorded correctly in the database, and are displayed as selected when I go into surveys to edit responses, but when I export to either pdf or xls it displays incorrect data.  What it seems to be doing is this - if 2 items were selected no matter which ones they are items 0 and 1 are displayed.. 1 item selected it shows item 0, regardless of which single item was selected..  So basically the logic seems to be x number of items were selected, so here I'll show you items 0 to x.

**Comment by meesir** [comment]
*0 to x-1 technically..

**Comment by kordan** [comment]
Thanks meesir
with your additional information I perfectly see and reproduce the issue.
I can confirm it.
I will start to look at it asap.
At the moment, thank you Brada65 and meesir both.

**Comment by kordan** [comment]
I have really no time to look at it seriously now.
I made a first draft of fix.
If you need it test it as far as you need.

Replace the switch from line 777 of  surveypro/field/multiselect/classes/field.php with:
```        // Output.
        // Here $answers is an array like: array(1,1,0,0).
        switch ($format) {
            case SURVEYPRO_ITEMSRETURNSVALUES:
                $answers = explode(SURVEYPRO_DBMULTICONTENTSEPARATOR, $content);
                $output = array();
                $values = $this->item_get_content_array(SURVEYPRO_VALUES, 'options');

                foreach ($answers as $answer) {
                    $output[] = $values[$answer];
                }
                $return = implode(SURVEYPRO_OUTPUTMULTICONTENTSEPARATOR, $output);
                break;
            case SURVEYPRO_ITEMRETURNSLABELS:
                $answers = explode(SURVEYPRO_DBMULTICONTENTSEPARATOR, $content);
                $output = array();
                $values = $this->item_get_content_array(SURVEYPRO_LABELS, 'options');

                foreach ($answers as $answer) {
                    $output[] = $values[$answer];
                }
                $return = implode(SURVEYPRO_OUTPUTMULTICONTENTSEPARATOR, $output);
                break;
            case SURVEYPRO_ITEMRETURNSPOSITION:
                // Here I will ALWAYS HAVE 0/1 so each separator is welcome, even ','.
                // I do not like pass the idea that ',' can be a separator so, I do not use it.
                $return = $content;
                break;
            default:
                $message = 'Unexpected $format = '.$format;
                debugging('Error at line '.__LINE__.' of '.__FILE__.'. '.$message , DEBUG_DEVELOPER);
        }
```
I can't understand now WHY this code seems to need to be different from the one of the checkbox... but I have too little time to think of it now.
Take it and let me know your opinion and test output, please.

Thanks.

**Comment by meesir** [comment]
I just gave it a try, same results..

**Comment by meesir** [comment]
I didn't do much testing, but this code gives the correct output for my use cases..
`        // Here $answers is an array like: array(1,1,0,0).
        switch ($format) {
            case SURVEYPRO_ITEMSRETURNSVALUES:
                $answers = explode(SURVEYPRO_DBMULTICONTENTSEPARATOR, $content);
                $output = array();
                $values = $this->item_get_content_array(SURVEYPRO_VALUES, 'options');

                
                foreach ($answers as $k) {                   
                        $output[] = $k;                   
                }
                $return = implode(SURVEYPRO_OUTPUTMULTICONTENTSEPARATOR, $output);
                break;
            case SURVEYPRO_ITEMRETURNSLABELS:
                $answers = explode(SURVEYPRO_DBMULTICONTENTSEPARATOR, $content);
                $output = array();
                $values = $this->item_get_content_array(SURVEYPRO_LABELS, 'options');

                
                foreach ($answers as $k) {                   
                        $output[] = $values[$k];                   
                }
                $return = implode(SURVEYPRO_OUTPUTMULTICONTENTSEPARATOR, $output);
                break;
            case SURVEYPRO_ITEMRETURNSPOSITION:
                // Here I will ALWAYS HAVE 0/1 so each separator is welcome, even ','.
                // I do not like pass the idea that ',' can be a separator so, I do not use it.
                $return = $content;
                break;
            default:
                $message = 'Unexpected $format = '.$format;
                debugging('Error at line '.__LINE__.' of '.__FILE__.'. '.$message , DEBUG_DEVELOPER);
        }

        return $return;
    }`

**Comment by meesir** [comment]
That's embarrassing, looking up, that's exactly what you have there.. I guess I didn't save after making the change or something, and have reinvented the wheel..  Either way, I guess the problem is solved..

**Comment by kordan** [comment]
Not sure I nderstood your words.
The code there was this:
https://github.com/kordan/moodle-mod_surveypro/blob/master/field/multiselect/classes/field.php#L777

**Comment by meesir** [comment]
After trying your fix it still didn't work for me, so I did some debugging and came up with my solution.  But as I compared to what you had offered as a fix I noticed my work was modified from the original code.  So I guess I failed to implement your fix properly and came up with a very similar solution to what you did myself.

**Comment by kordan** [comment]
Why do you use 
```
$output[] = $k;
```
in case SURVEYPRO_ITEMSRETURNSVALUES

and
```
$output[] = $values[$k];
```
in case SURVEYPRO_ITEMRETURNSLABELS?

**Comment by meesir** [comment]
I worked from your old code, and didn't change the iterator in the foreach..

**Comment by kordan** [comment]
yes, but the two snippets shouldn't be different

**Comment by meesir** [comment]
I figured one should return the index value and the other the label

**Comment by kordan** [comment]
I made few test during lunch time and my code was running fine in each test

**Comment by kordan** [comment]
No, as far as I remember the only difference is in $values.
In case SURVEYPRO_ITEMSRETURNSVALUES
$values holds values
In case SURVEYPRO_ITEMRETURNSLABELS
$values holds labels
All the other code doesn't change.

**Comment by meesir** [comment]
Right so in In case SURVEYPRO_ITEMSRETURNSVALUES the index would be returned which is what $k is.  So I probably should have also removed the line `$values = $this->item_get_content_array(SURVEYPRO_VALUES, 'options');` just before the loop, for it's not required.

**Comment by meesir** [comment]
Ohhh I see what you mean now.  My bad

**Comment by meesir** [comment]
I didn't test that aspect at all, I just made a quick assumption after fixing the label case..

**Comment by kordan** [comment]
Ok. Do your evaluations and let me know. Thanks a lot.

**Comment by Brada65** [comment]
I did test on local servers, everything is working like a charm. Will do later on production server and let both of you know what results are, but I think if it is working locally, that is that.

**Comment by kordan** [comment]
Nice to read, Vlad!

After the same feedback from Paul I will create one more branch in github.
Thank you both.

**Comment by kordan** [comment]
https://github.com/kordan/moodle-mod_surveypro/pull/342

**Comment by kordan** [comment]
As far as I know, the issue should be closed.
If nothing happen, I will close it as soon as #342 will be merged.
Thanks both.

**Comment by kordan** [comment]
As far as I know, this issue has been closed merging #342 just few minutes ago.
